pkg/client/googlegroup_github/mapper.go

```diff
@@ -19,15 +19,88 @@ import (
 	"context"
 	"fmt"
 	"os"
+	"slices"
 	"strings"
 
 	"google.golang.org/protobuf/encoding/prototext"
 
 	"github.com/abcxyz/pkg/logging"
 	"github.com/abcxyz/team-link/apis/v1alpha3"
+	"github.com/abcxyz/team-link/pkg/github"
 	"github.com/abcxyz/team-link/pkg/groupsync"
 )
 
+// GroupMapper implements groupsync.OneToManyGroupMapper
+// For now it was be used as Mapper between GoogleGroup
+// and GitHubTeam.
+// Both GoogleGroupMapper and GitHubMapper have similar
+// variables and functions, this way we don't need to repeatly
+// write similar function for both on them.
+type GroupMapper struct {
+	mappings map[string][]string
+}
+
+func (m *GroupMapper) AllGroupIDs(ctx context.Context) ([]string, error) {
+	res := make([]string, 0)
+	for key := range m.mappings {
+		res = append(res, key)
+	}
+	slices.Sort(res)
+	return res, nil
+}
+
+func (m GroupMapper) ContainsGroupID(ctx context.Context, key string) (bool, error) {
+	_, ok := m.mappings[key]
+	if !ok {
+		return false, fmt.Errorf("group %s is not mapped", key)
+	}
+	return ok, nil
+}
+
+func (m GroupMapper) MappedGroupIDs(ctx context.Context, key string) ([]string, error) {
+	x, ok := m.mappings[key]
+	if !ok {
+		return nil, fmt.Errorf("no mapping found for group ID: %s", key)
+	}
+	return x, nil
+}
+
+type GoogleGroupToGitHubMapper GroupMapper
+
+type GitHubToGoogleGroupMapper GroupMapper
+
+// NewBidirectionalGoogleGroupGitHubMapper creates a GoogleGroupToGitHubMapper
+// and a GitHubToGoogleGroupMapper using the provided groupMapping file.
+// Returns is (GoogleGroupToGitHubMapper, GitHubToGoogleGroupMapper, error).
+//
+// TODO: refactor this into client/googlegroup_github/mapper.go later.
+func NewBidirectionaGroupMapper(groupMappingFile string) (*GroupMapper, *GroupMapper, error) {
+	b, err := os.ReadFile(groupMappingFile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to read mapping file: %w", err)
+	}
+	tm := &v1alpha3.GoogleGroupToGitHubTeamMappings{}
+	if err := prototext.Unmarshal(b, tm); err != nil {
+		return nil, nil, fmt.Errorf("failed to unmarshal mapping file: %w", err)
+	}
+	ggToGHMapping := make(map[string][]string)
+	ghToGGMapping := make(map[string][]string)
+	for _, v := range tm.GetMappings() {
+		gitHubGroupID := github.Encode(v.GetGitHubTeam().GetOrgId(), v.GetGitHubTeam().GetTeamId())
+		if _, ok := ggToGHMapping[v.GetGoogleGroup().GetGroupId()]; !ok {
+			ggToGHMapping[v.GetGoogleGroup().GetGroupId()] = []string{gitHubGroupID}
+		} else {
+			ggToGHMapping[v.GetGoogleGroup().GetGroupId()] = append(ggToGHMapping[v.GetGoogleGroup().GetGroupId()], gitHubGroupID)
+		}
+		if _, ok := ghToGGMapping[gitHubGroupID]; !ok {
+			ghToGGMapping[gitHubGroupID] = []string{v.GetGoogleGroup().GetGroupId()}
+		} else {
+			ghToGGMapping[gitHubGroupID] = append(ghToGGMapping[gitHubGroupID], v.GetGoogleGroup().GetGroupId())
+		}
+	}
+	return &GroupMapper{mappings: ggToGHMapping}, &GroupMapper{mappings: ghToGGMapping}, nil
```

Comment:
I might be misunderstanding the goal here, but after reading the code, I'd like to understand why we think we need two different structures for this mapping? It seems like we're just taking the proto (which is flat) and merging in into `map[group][]users`? We should be able to manage this with a single data structure, since it seems like this is just creating a structure from the proto?

Reply:
These mappers will be used to construct a oneToManySyncer, which needs a sourceToDestMapper, and a destToSourceMapper. See(https://github.com/abcxyz/team-link/blob/main/pkg/groupsync/manytomanysyncer.go#L35) The [Sync](https://github.com/abcxyz/team-link/blob/main/pkg/groupsync/manytomanysyncer.go#L76) function needs mapping from both direction to handle the logics.

Reply:
Move the comment from the other thread (https://github.com/abcxyz/team-link/pull/78#discussion_r1872146301) here.

In this GoogleGroup to GitHub case, the structure is the same except the direction change, so use this `bidirectional` constructor, we can create both mappers with only one read from file and only one loop for the `mappings` to construct both mappers.

Reply:
Sorry, I'm still confused why this is two separate mappers? IIUC, we have a GitHub -> GoogleGroups and a GoogleGroups -> GitHub. Why can't that be a single structure that contains both maps?

Reply:
Sorry, I'm still confused why this is two separate mappers? IIUC, we have a GitHub -> GoogleGroups and a GoogleGroups -> GitHub. Why can't that be a single structure that contains both maps?

Reply:
Yes we can, both to create `NewManyToManySyncer`, but we still need to give it two mappers. See (https://github.com/abcxyz/team-link/blob/main/pkg/groupsync/manytomanysyncer.go#L46-L63).

But I think we can do 
```
type BidirectionGroupMapper struct {
  sourceMapper GroupMapper
  targetMapper GroupMapper
}
```

Reply:
Updated using the both method.